rplugin/python3/magma/images.py

```diff
@@ -130,12 +131,166 @@ def add_image(self, path: str, identifier: str, x: int, y: int, width: int, heig
+        self.height = height
+        self.nvim = nvim
+
+    def serialize_gr_command(self, **cmd):
+        payload = cmd.pop('payload', None)
+        cmd = ','.join('{}={}'.format(k, v) for k, v in cmd.items())
+        ans = []
+        w = ans.append
+        w(b'\033_G'), w(cmd.encode('ascii'))
+        if payload:
+            w(b';')
+            w(payload)
+        w(b'\033\\')
+        return b''.join(ans)
+
+
+    def write_chunked(self, **cmd):
+        from base64 import standard_b64encode
+
+        data = standard_b64encode(cmd.pop('data'))
+        while data:
+            chunk, data = data[:4096], data[4096:]
+            m = 1 if data else 0
+            self.nvim.lua.stdout.write(
+                self.serialize_gr_command(
+                    payload=chunk,
+                    m=m,
+                    **cmd
+                )
+            )
+            cmd.clear()
+
+
+    def show(self):
+        with open(self.path, 'rb') as f:
+            self.write_chunked(
+                a='T', # transmit directly to the terminal
+                i=self.id,
+                f=100, # for now, only png
+                v=self.height,
+                s=self.width,
+                C=1,
+                z=10,
+                q=2,
+                data=f.read(),
+            )
+
+    def hide(self):
+        self.nvim.lua.stdout.write(
+            self.serialize_gr_command(
+                i=self.id,
+                a='d', # remove image
+                q=2,
+            )
+        )
+
+
+class Kitty(Canvas):
+    nvim: Nvim
+    images: Dict[str, KittyImage]
+    to_make_visible: Set[str]
+    to_make_invisible: Set[str]
+    visible: Set[str]
+    next_id: int
+
+    def __init__(self, nvim):
+        self.nvim = nvim
+        self.images = {}
+        self.visible = set()
+        self.to_make_visible = set()
+        self.to_make_invisible = set()
+        self.next_id = 0
+        nvim.exec_lua("""
+            local fd = vim.loop.new_pipe(false)
+            fd:open(1)
+            local function write(data)
+                    fd:write(data)
+            end
+
+            stdout = {write = write}
+            """)
+
+    def init(self):
+        return self
+
+    def deinit(self):
+        return
+
+    def log(self,msg):
+        import pynvim
+        self.nvim.api.notify(
+            msg,
+            pynvim.logging.INFO,
+            {'title': "Magma"},
+        )
+
+    def present(self) -> None:
+        # images to both show and hide should be ignored
+        to_work_on = self.to_make_visible.difference(
+            self.to_make_visible.intersection(self.to_make_invisible))
+        self.to_make_invisible.difference_update(self.to_make_visible)
+        for identifier in self.to_make_invisible:
+            self.images[identifier].hide()
+            time.sleep(0.01)
+        for identifier in to_work_on:
+            image = self.images[identifier]
+            def fn(nvim, image):
```

Comment:
Pretty sure this function should receive `self` as well, since Python has some really weird "closures". (Then you can also do `self.nvim` instead of `nvim` as a param)

Reply:
this should be solved now as I don't use `self` in the callback.